### PR TITLE
IIS, buffer request body before taking lock

### PR DESCRIPTION
**Pull request by allanbomsft**
This is to avoid the situation where a slow transmitting client keeps the lock held. Buffer the entire request body outside of the main ModSec code, before the locks are taken.

### CODE REVIEW

**Comment by allanbomsft** [comment]
I have tested extensively on regular as well as chunked requests using my "rawrequestgen" test suite.

**Review by alonzop — COMMENTED** [review]
i just thought about the scenario in which we timeout due to not receiving all the body, are we freeing all the memory that was allocated?

**Review by alonzop — APPROVED** [review]
looks good

File: iis/mymodule.cpp [hunk 1]
```diff
@@ -752,6 +762,58 @@ CMyHttpModule::OnBeginRequest(
...
+        }
+        else {
+            // Append to linked list
+            cur->next = chunk;
+            cur = chunk;
+        }
+
+        if (hr != S_OK)
+        {
+            goto Finished;
```
Comment (puneetsaraswat) [inline iis/mymodule.cpp]: In this code path LeaveCriticalSection would be called twice. You would need to EnterCriticalSection before goto Finished.

File: iis/mymodule.cpp [hunk 2]
```diff
@@ -752,6 +762,58 @@ CMyHttpModule::OnBeginRequest(
...
+        DWORD readcnt;
+        HRESULT hr = pRequest->ReadEntityBody(chunk->data, HUGE_STRING_LEN, false, &readcnt, NULL);
+        if (ERROR_HANDLE_EOF == (hr & 0x0000FFFF))
+        {
+            free(chunk);
+            break;
+        }
+        chunk->length = readcnt;
+
+        if (!head)
```
Comment (puneetsaraswat) [inline iis/mymodule.cpp]: You can avoid this special case for head. How about this:
preAllocBodyChunk** cur = &head;

<snip><snip>

*cur = chunk;
cur = &(chunk->next);

File: iis/mymodule.cpp (hunk 1 above)
Comment (alonzop) [reply iis/mymodule.cpp]: another option will be to add another goto right after the release of the lock

Comment (allanbomsft) [reply iis/mymodule.cpp]: refactored into a func without the goto, so now always re-entering crit sec

Comment (alonzop) [reply iis/mymodule.cpp]: why not save the time of the lock/unlock if it is not needed?

Comment (allanbomsft) [reply iis/mymodule.cpp]: ok, added extra goto label

**Review by puneetsaraswat — APPROVED** [review]

**Pull request merged**